accepted/2022/Multiple-Equivalent-Framework-Support-TFM-As-Aliases.md

```diff
@@ -0,0 +1,397 @@
+
+### More significant assets file schema changes?
+
+Should we consider more significant schema changes in the assets file?
+
+Under `$/libraries`, each package lists every file in the package.
+However, the performance penalty of needing to parse that list every time the assets file is read may be worse than any benefit it provides from avoiding enumerating the filesystem when it is needed.
+The assets file primary purpose is to tell the .NET SDK and dotnet/NuGet.BuildTools what package assets are used by the project, and they don't need or use the `$/libraries` section at all.
+While reading the assets file might be a small percentage of the overall build process, if we consider reading the assets file in isolation, there could be a fairly significant performance increase by removing this entire section of the assets file.
+
+`$/projectFileDependencyGroups` appears to be a duplication of the information available in `$/project/frameworks`.
+
+`$/project/originalTargetFrameworks` duplicates information already available in `$/project/restore/frameworks`
+
+Since the .NET SDK has to read the assets file after NuGet writes it, time and memory allocations spent on parts of the assets file not relevant to asset selection reduces performance.
+Additionally, MSBuild binlogs embed the assets file, which bloats the binlog size.
+
+Removing the duplicated data should be non-controversial.
+But the `$/libraries` section is useful for diagnosing customer complaints.
+Whether impacting the performance of every build is worth occasional benefit in helping debug customer issues is subjective.
+An alternative is we add a `<RestoreVerboseAssetsFile>` property to include it, but otherwise have it off by default for performance reasons.
+Although the .NET SDK already have a cache, so they should only read the assets file when the assets file changes.
+
+### More significant lock file schema changes
+
+I don't want to make the scope of this feature so large that it takes an unreasonably long time to implement.
+However, breaking changes are rarely made in NuGet, so if we have to make a breaking change to enable a feature, it's tempting to take advantage and propose other changes at the same time.
+
+Since the lock file contains the package's content hash, and the package's `nuspec` is included in that content hash, this means that the dependencies in the the lock file are a redundant duplication of information available elsewhere.
+
+## Future Possibilities
+
+<!-- What future possibilities can you think of that this proposal would help with? -->
+
+### Extending Pack
+
+It seems feasible that if a single project can target multiple versions of Visual Studio, through a new Visual Studio Extensibility SDK, that someone will also want to create a package that supports multiple versions of Visual Studio.
+
+Designing how this could work is out of scope for this spec, and can be introduced in a different spec.
```

Comment:
Just writing down what we discussed in yesterday's meeting. For runtime to not be broken by this change, we need to keep supporting what works today.

In runtime, we set `IncludeBuildOutput=false` and `SuppressDependenciesWhenPacking=true` for custom TFMs (one that use a TPM not known by the SDK -> i.e. `net8.0-unix`).

Now if NuGet starts emitting an error for TFM aliases it would break runtime. It would be good to only emit an error when the TFM is included in the package (either the build output or the dependencies section). So in technical terms, only emit an error when `TFM == alias && (IncludeBuildOutput == true or SuppressDependenciesWhenPackaging == false)`